tests/utils.py

```diff
@@ -154,69 +161,169 @@ class TestServerProcess():
+    """
+    Worker function to run in a seprate thread.
+    Reads from 'stream' (stdout), puts lines in Queue (Queue is thread-safe).
+    """
+
+    while True:
+      # readline() is a blocking operation;
+      # We need to decode because readline returns a sequence of 8-bit bytes,
+      # but we want to put the messages in the queue as strings.
+      log_line = stream.readline().decode()
+      line_queue.put(log_line)
+
+      if len(log_line) == 0:
+        # This mean we are at the end of the stream meaning the server process
+        # has exited and we don't need the stream anymore.
+        stream.close()
+        break
+
+
+
+  def _set_port_if_in_logs(self, timeout):
+    """
+    Checks if the port is logged from the server subprocess.
+    If it's found, self.port is set and returns True.
+    Otherwise, returns False.
+    TestServerProcessError is throwned if the process has exited
+    or if it has failed to bind.
+    """
+
+    # We have hardcoded the message we expect on a successful server startup.
+    expected_msg = 'bind succeeded, server port is: '
+    # Get lines until port is found, child process exits, or we timeout.
+    try:
+      line = self._log_queue.get(timeout=timeout)
+      if len(line) == 0:
+        # The process has exited.
+        raise TestServerProcessError(self.server + ' exited unexpectedly ' \
+            + 'with code ' + str(self.__server_process.poll()) + '!')
+
+      elif line.startswith(expected_msg):
+        self.port = int(line[len(expected_msg):])
+        return True
+      else:
+        # An exception or some other message is printed from the server.
+        self.__logged_messages.append(line)
+        # Check if more lines are logged.
         self.flush_log()
-        raise e
+        raise TestServerProcessError(self.server + ' failed to bind!')
+    except queue.Empty:
+      self.__logger.info("No log lines are found in the Queue!")
+      return False
 
 
 
-  def flush_log(self):
-    """Logs contents from TempFile, truncates buffer"""
+  def _kill_server_process(self):
+    """Kills the server subprocess if it's running."""
 
-    # Seek is needed to move the pointer to the beginning of the file, because
-    # the subprocess could have read and/or write and thus moved the pointer.
-    self.__temp_log_file.seek(0)
-    log_message = self.__temp_log_file.read()
+    if self.is_process_running():
+      self.__logger.info('Server process ' + str(self.__server_process.pid) +
+          ' terminated.')
+      self.__server_process.kill()
+      self.__server_process.wait()
 
-    if len(log_message) > 0:
+
+
+  def flush_log(self):
+    """Flushes the log lines from the logging queue."""
+
+    while True:
+      # Get lines from log_queue
+      try:
+        line = self._log_queue.get(block=False)
+        if len(line) == 0:
+          # This means no more lines will be sent in the queue.
+          break
+        else:
+          self.__logged_messages.append(line)
```

Comment:
```suggestion
        if len(line) > 0:
          self.__logged_messages.append(line)
```
Just a suggestion to make it easier to read: the break should not be required, next get() raises anyway